### PR TITLE
ASoC: SOF: refine sof for multi-channels support

### PR DESCRIPTION
**Pull request by RanderWang**
User mode lib calls info function in kcontrol to get the channel number of
volume mixer. The default info function is snd_soc_info_volsw which only
supports max two channels. Just like get/put function,  info function is
bound to a bespoke info function in SOF and return correct channel number
to user.

The snd_sof_volume_info is based on snd_soc_info_volsw, and refines the
channel setting. Using standard SND_SOC_TPLG_CTL_VOLSW to register our
kcontrol function because only it supports binding all kcontrol functions to bespoke
ones. And this change doesn't affect the volume mixer which is not created by
topology and doesn't affect the topology released

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/control.c [hunk 1]
```diff
@@ -98,6 +98,31 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
...
+	struct snd_sof_control *scontrol = mc->dobj.private;
+	unsigned int channels = scontrol->num_channels;
+	int platform_max;
+
+	if (!mc->platform_max)
+		mc->platform_max = mc->max;
+
+	platform_max = mc->platform_max;
+
+	if (platform_max == 1 && !strstr(kcontrol->id.name, " Volume"))
```
Comment (lgirdwood) [inline sound/soc/sof/control.c]: Cant do this string checking, best to have separate info() calls for volume and switch

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -1918,6 +1918,8 @@ static int sof_manifest(struct snd_soc_component *scomp, int index,
 
 /* vendor specific kcontrol handlers available for binding */
 static const struct snd_soc_tplg_kcontrol_ops sof_io_ops[] = {
+	{SND_SOC_TPLG_CTL_VOLSW, snd_sof_volume_get, snd_sof_volume_put,
+		snd_sof_volume_info},
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c]: Should this info be appended to SOF_TPLG_KCTL_VOL_ID ?

Comment (RanderWang) [reply sound/soc/sof/topology.c]: yes, I would do this.  I did this according to id, the smallest is in the head

File: sound/soc/sof/control.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/sof/control.c]: Yes, I will refine it. And I also checked the Asoc, the info function would be bound only for  volume case

**Comment by RanderWang** [comment]
Update my comments and patch

File: sound/soc/sof/topology.c [hunk 3]
```diff
@@ -244,7 +244,7 @@ static int sof_control_load_volume(struct snd_soc_component *scomp,
 	struct sof_ipc_ctrl_data *cdata;
 
 	/* validate topology data */
-	if (le32_to_cpu(mc->num_channels) >= SND_SOC_TPLG_MAX_CHAN)
+	if (le32_to_cpu(mc->num_channels) > SND_SOC_TPLG_MAX_CHAN)
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c]: This should be a separate fix in it's own PR.

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -1919,6 +1919,8 @@ static int sof_manifest(struct snd_soc_component *scomp, int index,
 /* vendor specific kcontrol handlers available for binding */
 static const struct snd_soc_tplg_kcontrol_ops sof_io_ops[] = {
 	{SOF_TPLG_KCTL_VOL_ID, snd_sof_volume_get, snd_sof_volume_put},
+	{SND_SOC_TPLG_CTL_VOLSW, snd_sof_volume_get, snd_sof_volume_put,
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c:1922]: Still not following why we need to override a standard kcontrol ID here.

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: ok, I will make another PR

Comment (plbossart) [reply sound/soc/sof/topology.c]: Can  you rebase and repush, I don't see why this change is still visible.

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:1922]: I also don't get what this is about?

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: Sorry, I will do it

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:1922]: @plbossart  @lgirdwood  please refer to https://github.com/thesofproject/soft/pull/34

**Comment by RanderWang** [comment]
Refine my patch.  The detail of this change is discussing in https://github.com/thesofproject/soft/pull/34
for short, I add SND_SOC_TPLG_CTL_VOLSW to bind get & put & info function to some bespoke ones.  The info function can't be bound to a bespoke one with SOF_TPLG_KCTL_VOL_ID.
